cubi_tk/sodar/update_samplehseet.py

```diff
@@ -0,0 +1,628 @@
+        overwrite: bool = False,
+        no_autofill: bool = False,
+    ):
+        if not all(update_table.columns.isin(isa_table.columns)):
+            raise ValueError(
+                "New ISA table has columns that are not present in the existing ISA table."
+            )
+
+        # reorder update_table columns to match isa_table
+        update_table = update_table[
+            [col for col in isa_table.columns if col in update_table.columns]
+        ]
+
+        # check for matches with existing data based on required cols (which should be the material cols)
+        # separate new data into updates (all mat_cols match) and additions
+        mat_cols = [
+            col
+            for col in isa_table.columns
+            if orig_col_name(col) in REQUIRED_COLUMNS + REQUIRED_IF_EXISTING_COLUMNS
+        ]
+        common_rows_update = (
+            update_table[mat_cols].isin(isa_table[mat_cols].to_dict(orient="list")).all(axis=1)
+        )
+        common_rows_isa = (
+            isa_table[mat_cols].isin(update_table[mat_cols].to_dict(orient="list")).all(axis=1)
+        )
+        updates = update_table.loc[common_rows_update].copy()
+        additions = update_table.loc[~common_rows_update].copy()
+
+        if overwrite:
+            # update all isa_table columns with new_data
+            isa_table.loc[common_rows_isa, updates.columns] = updates.values
+        else:
+            # Update only values those columns that are empty/falsy
+            # Give a warning if any existing values are different
+            for col in updates.columns:
+                isa_col = isa_table[col].loc[common_rows_isa]
+                equal_values = isa_col.reset_index(drop=True) == updates[col].reset_index(drop=True)
+                empty_values = isa_col.isnull().reset_index(drop=True) | isa_col.eq("").reset_index(
+                    drop=True
+                )
+                clash_rows = ~equal_values & ~empty_values
+
+                if equal_values.all():
+                    continue
+                elif clash_rows.any():
+                    clash = isa_table.loc[isa_col.index[clash_rows], mat_cols + [col]]
+                    clash["<New values:>" + col] = updates[col].loc[clash_rows].values
+                    logger.warning(
+                        f"Given values for ISA column '{col}' have different existing values, "
+                        "these will not be updated. Use `--overwrite` to force update.\n"
+                        + clash.to_string(index=False, na_rep="")
+                    )
+
+                isa_table.loc[isa_col.index[empty_values], col] = (
+                    updates[col].loc[empty_values].values
+                )
+
+        # Check which cols should be autofilled ("Protocol REF" needs to be, for ISA to work)
+        autofill_cols = {
+            col: isa_table[col].unique()[0]
+            for col in isa_table.columns
+            if orig_col_name(col) == "Protocol REF"
```

Comment:
So _only_ columns named "Protocol REF" will be autofilled?

Reply:
Yes, I initially wanted to (always) autofill some other columns as well, but that was able to break ISA compatibility.

Reply:
Ah okay. Then this is a bit convoluted for the singular column that is auto filled, but might become useful in the future?